### PR TITLE
 Internalize Ownership Management of Volumes into Plugins

### PR DESCRIPTION
**Pull request by ghost**
This is an implementation of the proposal merged in #18521

The assigned fsGroup is passed on to the individual volume plugins from the kubelet. The individual volume plugins set the ownership of their volumes where desired.

@pmorie

### CODE REVIEW

**Comment by swagiaal** [comment]
@k8s-bot test this.

File: pkg/volume/aws_ebs/aws_ebs_test.go [hunk 1]
```diff
@@ -154,7 +154,7 @@ func TestPlugin(t *testing.T) {
 		t.Errorf("Got unexpected path: %s", path)
 	}
 
-	if err := builder.SetUp(); err != nil {
+	if err := builder.SetUp(nil); err != nil {
```
Comment (pmorie) [inline pkg/volume/aws_ebs/aws_ebs_test.go]: add a check for fsgroup in the unit tests?

**Comment by swagiaal** [comment]
@k8s-bot unit test this

File: pkg/volume/aws_ebs/aws_ebs.go [hunk 2]
```diff
@@ -291,6 +290,10 @@ func (b *awsElasticBlockStoreBuilder) SetUpAt(dir string) error {
 		return err
 	}
 
+	if !b.readOnly {
```
Comment (pmorie) [inline pkg/volume/aws_ebs/aws_ebs.go]: @thockin I don't think we settled on the impl detail of whether you make this call if fsGroup == 0, thoughts?

Comment (thockin) [inline pkg/volume/aws_ebs/aws_ebs.go]: We need to ensure the post-condition.  If it is safe to assume that kubelet's fsgid is 0, then we can elide this call - is that a safe assumption?  I am not sure.

Comment (swagiaal) [inline pkg/volume/aws_ebs/aws_ebs.go]: I don't quite follow... Why do we need special handling for fsGroup == 0. Please note that fsGroup is an *int64 for which the unset case is fsGroup == nil

Comment (thockin) [inline pkg/volume/aws_ebs/aws_ebs.go]: My point was that the post-condition is that ownership is set.  If fsGroup is 0, we need to ensure that everything is gid=0.  That *should* be the default today, but I don't know if it is safe to assume it always will be.

Comment (swagiaal) [inline pkg/volume/aws_ebs/aws_ebs.go]: Ah I see... Yeah I don't know if we can assume that the default ownership will always be 0. In addition (for better or for worse) the ownership management also sets `setgid`, and that is less likely to be the default so we might as well run it.

File: pkg/volume/aws_ebs/aws_ebs_test.go (hunk 1 above)
Comment (swagiaal) [inline pkg/volume/aws_ebs/aws_ebs_test.go]: So it looks like that would not be possible to do. The problem is that you cannot chown as non-root. This can be unit tested though by injecting chmodRunner, chownRunner but I think that should be a separate PR as this one is pretty large already.

File: pkg/volume/aws_ebs/aws_ebs.go [hunk 3]
```diff
@@ -222,20 +222,19 @@ var _ volume.Builder = &awsElasticBlockStoreBuilder{}
 
 func (b *awsElasticBlockStoreBuilder) GetAttributes() volume.Attributes {
 	return volume.Attributes{
-		ReadOnly:                    b.readOnly,
-		Managed:                     !b.readOnly,
-		SupportsOwnershipManagement: true,
-		SupportsSELinux:             true,
+		ReadOnly:        b.readOnly,
+		Managed:         !b.readOnly,
```
Comment (pmorie) [inline pkg/volume/aws_ebs/aws_ebs.go]: I don't think we need the `Managed` field of `volume.Attributes` anymore - remove it?

**Comment by pmorie** [comment]
@swagiaal Have you tested this with any non-emptyDir volumes?

File: pkg/volume/aws_ebs/aws_ebs_test.go (hunk 1 above)
Comment (pmorie) [inline pkg/volume/aws_ebs/aws_ebs_test.go]: You're right.

**Comment by swagiaal** [comment]
> @swagiaal Have you tested this with any non-emptyDir volumes?

Yes. I have tested Cinder, iSCSI, and RBD I also added a check to the e2e tests and ran them. See #19612. 

I also have a branch adding the check to AWS EBS, and GCE PD but that one is blocked on enabling security context in the e2e test suite, but it is on my list

**Comment by swagiaal** [comment]
hmm.. this has been all green for a while but the bot has not gotten to it

**Comment by pmorie** [comment]
The merge queue has been blocked a lot recently due to test flakes.  The
best way to help is to check out the flake issues and try to bust one.
